### PR TITLE
add PuppetDB configuration on puppet master side

### PR DESCRIPTION
**Pull request by mmoll**
more a WIP for discussion. comments please :newspaper:

### CODE REVIEW

**Comment by mmoll** [comment]
rebased

**Comment by ekohl** [comment]
I wouldn't be against this, but I'm hesitant to add it as an unconditional dependency. I'm leaning to doing so because IMHO PuppetDB is a valuable addition in many cases.

**Comment by mmoll** [comment]
now with added test. no idea what's going on with 2.7 there...

**Comment by ekohl** [comment]
I wonder if a gem broke compatibility with 2.7.

**Comment by mmoll** [comment]
I messed up metadata.json. All :green_heart: now

**Comment by ekohl** [comment]
Maybe add a test to the 'with no customer parameters' that it should not contain the puppetdb class?

Other than that there's a puppetlabs recommendation that for soft dependencies it should be in README rather than metadata.json. See https://github.com/theforeman/puppet-puppet/pull/162#issuecomment-72476020 as well.

**Comment by mmoll** [comment]
all done, I think

**Comment by ekohl** [comment]
:+1: from my side. @domcleal?

**Comment by mmoll** [comment]
needed a rebase again

File: README.md [hunk 1]
```diff
@@ -71,6 +72,15 @@ wrapper classes or even your ENC (if it supports param classes). For example:
       server_external_nodes => '',
     }
 
+    # Want to integrate with an existing PuppetDB?
+    class { '::puppet':
+      server               => true,
+      server_puppetdb_host => 'mypuppetdb.example.com',
+      server_reports       => 'puppetdb',
```
Comment (domcleal) [inline README.md]: Nitpick, should we keep 'foreman' in here too?

File: manifests/init.pp [hunk 2]
```diff
@@ -415,6 +423,9 @@
   $server_foreman_ssl_key        = $puppet::params::client_ssl_key,
   $server_facts                  = $puppet::params::server_facts,
   $server_puppet_basedir         = $puppet::params::server_puppet_basedir,
+  $server_puppetdb_host          = $puppet::params::server_puppetdb_host,
```
Comment (domcleal) [inline manifests/init.pp]: validate_string on this?

File: manifests/init.pp [hunk 3]
```diff
@@ -415,6 +423,9 @@
...
+  $server_puppetdb_port          = $puppet::params::server_puppetdb_port,
```
Comment (domcleal) [inline manifests/init.pp]: validate_int?

File: manifests/init.pp (hunk 2 above)
Comment (domcleal) [inline manifests/init.pp]: Oh no, it's `undef`

**Comment by domcleal** [comment]
Otherwise :+1:

File: manifests/init.pp (hunk 2 above)
Comment (ekohl) [inline manifests/init.pp]: We don't have it now, but what we could do:
```puppet
if $server_puppetdb_host {
  validate_string($server_puppetdb_host)
}
```

File: manifests/init.pp (hunk 3 above)
Comment (ekohl) [inline manifests/init.pp]: That doesn't exist afaik

File: manifests/init.pp (hunk 2 above)
Comment (domcleal) [inline manifests/init.pp]: Good idea.

File: manifests/init.pp (hunk 3 above)
Comment (domcleal) [inline manifests/init.pp]: You're right...

File: README.md (hunk 1 above)
Comment (mmoll) [inline README.md]: As foreman got optional, I presume not, but I wouldn't mind having it also in just as example.

**Comment by mmoll** [comment]
updated

**Comment by domcleal** [comment]
Merged, thanks @mmoll!

File: README.md (hunk 1 above)
Comment (ekohl) [inline README.md]: I do think people will be surprised if suddenly they don't have foreman reports anymore so I'd prefer having it in the example.

Comment (ekohl) [inline README.md]: And just now I see it was already merged.

Comment (domcleal) [inline README.md]: Ah, sorry.  Please feel free to push it straight to the repo, I guess it'd be `server_reports => 'puppetdb,foreman'`

Comment (ekohl) [inline README.md]: No problem, it's just a minor comment :)

Comment (ekohl) [inline README.md]: Fixed in d640db097b39531c5385ccfcffdb28ad4f405d89

**Pull request merged**